<a href="https://colab.research.google.com/github/alirezakhaleghianaghizi/ml_1401/blob/main/hw3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

علیرضا خالقی آناقیزی 
99101462

سوال 2

در قسمت زیر فایل های مورد نظر را در گوگل درایو ذخیره کردیم و سپسفایل های مورد نظر را لود کرده و دیتا های 
موجود را به ترتیب در ماتریس های 

SVT 

ذخیره کردیم.

In [2]:
import csv 
import numpy as np
from google.colab import drive

drive.mount('/content/drive')
filename1='/content/drive/MyDrive/ml/fashion-mnist.csv' 

fields = []
data = []
ValidationSet=[]
with open(filename1, 'r') as csvfile:
    # creating a csv reader object
    csvreader = csv.reader(csvfile)
     
    # extracting field names through first row
    fields = next(csvreader)
 
    # extracting each data row one by one 
    for row in csvreader:
      data.append(row)  

data=np.array(data,dtype=int).T
#print(data)
dataX=np.array(data[1:len(data)-1]).T
dataY=data[len(data)-1]
print(len(dataX))
print(len(dataY))

Mounted at /content/drive
10000
10000


در قسمت زیر با استفاده از کتابخانه 

sklearn 

با استفاده از تابع جدا سازی ابتدا ده ها ی نمرین را جدا کردیم و سپس از باقیمانده داده ها ولیدیشن و تست را جدا کردیم.

In [3]:
from sklearn.model_selection import train_test_split
train_ratio = 0.50
validation_ratio = 0.35
test_ratio = 0.15

# train is now 50% of the entire data set
x_train, x_test, y_train, y_test = train_test_split(dataX, dataY, test_size=1 - train_ratio)

# test is now 10% of the initial data set
# validation is now 15% of the initial data set
x_val, x_test, y_val, y_test = train_test_split(x_test, y_test, test_size=test_ratio) 

print(x_train, x_val, x_test)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]] [[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]] [[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


روش اول :
با استفاده از کتابخانه 

sklearn

توابع اس وی ام را خارج کردیم و از تابع خطی آن یادگیری را انجام دادیم.

In [6]:
from sklearn import svm
lin_clf = svm.LinearSVC()
lin_clf.fit(x_train,y_train)
y_val_pred=lin_clf.predict(x_val)
confusion_val_linear=get_confusion(y_pred=y_val_pred,y_true=y_val)
print(confusion_val_linear)
A_val_linear=np.trace(confusion_val_linear)/np.sum(confusion_val_linear)
print(A_val_linear)

/usr/local/lib/python3.8/dist-packages/sklearn/svm/_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


[[285, 4, 14, 27, 5, 0, 61, 2, 5, 0], [5, 394, 6, 11, 1, 0, 3, 3, 0, 0], [19, 3, 298, 3, 58, 0, 52, 1, 3, 0], [22, 11, 24, 314, 30, 0, 18, 0, 2, 3], [6, 1, 58, 14, 249, 0, 83, 0, 3, 0], [4, 0, 6, 3, 1, 328, 5, 45, 7, 13], [65, 2, 63, 14, 40, 0, 233, 0, 4, 0], [0, 0, 1, 1, 0, 17, 2, 394, 0, 35], [12, 1, 18, 11, 13, 12, 17, 3, 349, 0], [1, 1, 0, 0, 0, 11, 1, 20, 2, 394]]
0.7618823529411765


در قسمت زیر تابعی مینویسم که با گرفتن لیبل های پیش بینی شده و لیبل های اصلی ماتریس خطا را میسازیم

In [5]:
def get_confusion(y_pred,y_true):
  confusion=[]
  for j in range(10):
    row=[]
    for i in range(10):
     k=0;
     for m in range(len(y_pred)):
       if y_pred[m]==i :
         if y_true[m]==j:
           k=k+1
     row.append(k)    
    confusion.append(row)
  return confusion

در این روش چون ما پارامتر انتخابی نداریم همان پیش بینی را برای داده های تست انجام میدهیم و ماتریس مورد نظر و دقت آن را مینویسیم.

In [7]:
y_test_pred=lin_clf.predict(x_test)
confusion_test_linear=get_confusion(y_pred=y_test_pred,y_true=y_test)
print(confusion_test_linear)
A_test_linear=np.trace(confusion_test_linear)/np.sum(confusion_test_linear)
print(A_test_linear)

[[49, 1, 2, 3, 2, 0, 7, 0, 1, 0], [2, 71, 2, 1, 0, 0, 1, 0, 0, 0], [4, 0, 57, 3, 8, 1, 8, 1, 0, 0], [7, 1, 6, 56, 4, 0, 4, 0, 0, 0], [1, 1, 9, 4, 50, 0, 18, 0, 1, 0], [0, 0, 0, 0, 0, 46, 0, 4, 2, 2], [8, 0, 12, 4, 9, 0, 46, 0, 3, 0], [0, 0, 0, 0, 0, 4, 0, 69, 0, 1], [3, 0, 4, 1, 1, 1, 8, 1, 71, 0], [0, 0, 0, 0, 0, 1, 0, 7, 0, 56]]
0.7613333333333333


دیدیم در هر دو داده های ولیدیشن وتست دقت ما در حدود 0.75 میباشد و ماتریس های خطا نیز در بالا قابل مشاهده هستند وچون داده های ولیدیشن از دو برابر تست بیشتر است ماتریس آن نیز درایه هایش بزرگتر هستند.

روش دوم :
در اینجا ما کرنل مورد نظر را به صورت گاووسی قرار میدهیم.
گوسی همان 

RBF

با این تقاوت که پارامتر 

c

 آن برابر یک در نظر گرفته میشود . 
 حال برای یک بازه از مقدار گاما میبینیم بهترین دقت برای کدام یکی است .
 

In [9]:
from sklearn.svm import SVC
gamma_1d_range = np.logspace(-9,3,13)
classifiers = []
min_gamma=0;
max_A=0;
for mygamma in gamma_1d_range:
  clf = SVC( kernel='rbf',C=1, gamma=mygamma)
  clf.fit(x_train, y_train)
  y_val_pred=clf.predict(x_val)
  confusion_val_gaussian=get_confusion(y_pred=y_val_pred,y_true=y_val)
  #print(confusion_val_gaussian)
  A_val_gaussian=np.trace(confusion_val_gaussian)/np.sum(confusion_val_gaussian)
  classifiers.append(A_val_gaussian)
  if A_val_gaussian>max_A:
    min_gamma=mygamma
    max_A=A_val_gaussian
print(min_gamma)
print(classifiers)



        

1e-07
[0.6581176470588235, 0.772235294117647, 0.8435294117647059, 0.8374117647058823, 0.18305882352941177, 0.09482352941176471, 0.09482352941176471, 0.09482352941176471, 0.09482352941176471, 0.09482352941176471, 0.09482352941176471, 0.09482352941176471, 0.09482352941176471]


طبق محاسبات بالا برای گاما در حدود ده به توان منفی 7 مقدار دقت ما بیشینه شده است 
حال در حدود این اوردر دوباره نگاه نگاه میکنه ببینیم چه ضریبی در این اوردر مناسب است 
نکته محاسبات بالا در حدود بیست دقیقه طول کشید و یکمقدار کند بوده است دوباره بیست دقیقه لازم است تا ضریب را به دست آوریم

In [11]:
from sklearn.svm import SVC
gamma_2d_range = (np.array(range(9))+1)*min_gamma
classifiers = []
conf_Matrixes=[]
min_gamma=0;
max_A=0;
for mygamma in gamma_2d_range:
  clf = SVC( kernel='rbf',C=1, gamma=mygamma)
  clf.fit(x_train, y_train)
  y_val_pred=clf.predict(x_val)
  confusion_val_gaussian=get_confusion(y_pred=y_val_pred,y_true=y_val)
  #print(confusion_val_gaussian)
  A_val_gaussian=np.trace(confusion_val_gaussian)/np.sum(confusion_val_gaussian)
  classifiers.append(A_val_gaussian)
  if A_val_gaussian>max_A:
    min_gamma=mygamma
    max_A=A_val_gaussian
    conf_Matrixes=confusion_val_gaussian
print(min_gamma)
print(classifiers)
print(conf_Matrixes)

4e-07
[0.8435294117647059, 0.8541176470588235, 0.8569411764705882, 0.8588235294117647, 0.8564705882352941, 0.8550588235294118, 0.852, 0.8494117647058823, 0.8451764705882353]
[[342, 1, 5, 14, 0, 0, 37, 0, 4, 0], [1, 398, 8, 13, 1, 0, 2, 0, 0, 0], [4, 0, 346, 5, 41, 0, 37, 0, 4, 0], [24, 1, 4, 374, 14, 0, 7, 0, 0, 0], [0, 0, 31, 6, 337, 0, 38, 0, 2, 0], [1, 0, 0, 0, 0, 377, 0, 18, 5, 11], [74, 0, 60, 6, 27, 0, 251, 0, 3, 0], [0, 0, 0, 0, 0, 11, 0, 415, 0, 24], [2, 0, 3, 1, 4, 5, 7, 3, 411, 0], [0, 0, 0, 0, 0, 9, 0, 21, 1, 399]]


با توجه به نتایج بالا بهترین گاما برابر 4در ده به توان منفی 7 هستش که دقت آن برابر 0.8588 میباشد 
حال با توجه به این مقدار برای داده های تست دقت و ماتریس را تشکیل میدهیم.

In [12]:
from sklearn.svm import SVC
best_gamma_gaussian=min_gamma
clf = SVC( kernel='rbf',C=1, gamma=best_gamma_gaussian)
clf.fit(x_train, y_train)
y_test_pred=clf.predict(x_test)
confusion_test_gaussian=get_confusion(y_pred=y_test_pred,y_true=y_test)
print(confusion_test_gaussian)
A_test_gaussian=np.trace(confusion_test_gaussian)/np.sum(confusion_test_gaussian)
print(A_test_gaussian)

[[57, 0, 1, 1, 0, 0, 5, 0, 1, 0], [1, 71, 1, 3, 0, 0, 0, 0, 1, 0], [2, 0, 63, 1, 12, 0, 4, 0, 0, 0], [4, 0, 1, 68, 4, 0, 1, 0, 0, 0], [0, 0, 6, 1, 64, 0, 13, 0, 0, 0], [0, 0, 0, 0, 0, 50, 0, 2, 1, 1], [13, 0, 7, 0, 8, 0, 50, 0, 4, 0], [0, 0, 0, 0, 0, 1, 0, 69, 0, 4], [0, 0, 2, 0, 0, 1, 3, 2, 82, 0], [0, 0, 0, 0, 0, 0, 0, 3, 0, 61]]
0.8466666666666667


نتیجه نهایی در کرنل گووسی با گاما مورد نظر دقت برابر 0.846 و ماتریس مورد نظر نیز در بالا میباشد که نسبت به یادگیری خطی نتیجه بهتری به دست آوردیم .

In [ ]:
#from sklearn.gaussian_process import GaussianProcessClassifier
#from sklearn.gaussian_process.kernels import RBF
#classifiers = []
#gammas=np.logspace(-9,3,13)
#min_gamma=0;
#max_A=0;
#for gamma in gammas :
#  kernel = 1.0 * RBF(gamma)
#  gpc = GaussianProcessClassifier(kernel=kernel, random_state=0).fit(x_train, y_train)
#  y_val_pred=gpc.predict(x_val)
#  confusion_val_gaussian=get_confusion(y_pred=y_val_pred,y_true=y_val)
  #print(confusion_val_gaussian)
#  A_val_gaussian=np.trace(confusion_val_gaussian)/np.sum(confusion_val_gaussian)
#  classifiers.append(A_val_gaussian)
#  if A_val_gaussian>max_A:
#    min_gamma=mygamma
#    max_A=A_val_gaussian
#print(min_gamma)
#print(classifiers)